File: 2023/18/solution.js
const fs = require("fs");
const path = require("path");
const MinHeap = require("../../utils/MinHeap");

const input = fs
  .readFileSync(path.join(__dirname, "input.txt"), "utf-8")
  .split("\n");

class Instruction {
  constructor(line) {
    const [a, b, c] = line.split(" ");
    this.dir = a;
    this.dist = Number(b);
    this.color = c.slice(1, -1);
  }
}

const pos = [0, 0];
const instructions = input.map((line) => new Instruction(line));
let minX = 0;
let minY = 0;
let maxX = 0;
let maxY = 0;
for (const instruction of instructions) {
  switch (instruction.dir) {
    case "U":
      pos[1] -= instruction.dist;
      break;
    case "D":
      pos[1] += instruction.dist;
      break;
    case "R":
      pos[0] += instruction.dist;
      break;
    case "L":
      pos[0] -= instruction.dist;
      break;
  }
  minX = Math.min(minX, pos[0]);
  minY = Math.min(minY, pos[1]);
  maxX = Math.max(maxX, pos[0]);
  maxY = Math.max(maxY, pos[1]);
}

const width = maxX - minX + 1;
const height = maxY - minY + 1;
const grid = Array(height)
  .fill(0)
  .map(() => Array(width).fill("."));

[pos[0], pos[1]] = [0 - minX, 0 - minY];

for (const instruction of instructions) {
  switch (instruction.dir) {
    case "U":
      for (let i = 0; i < instruction.dist; i++) {
        pos[1]--;
        grid[pos[1]][pos[0]] = "#";
      }
      break;
    case "D":
      for (let i = 0; i < instruction.dist; i++) {
        pos[1]++;
        grid[pos[1]][pos[0]] = "#";
      }
      break;
    case "R":
      for (let i = 0; i < instruction.dist; i++) {
        pos[0]++;
        grid[pos[1]][pos[0]] = "#";
      }
      break;
    case "L":
      for (let i = 0; i < instruction.dist; i++) {
        pos[0]--;
        grid[pos[1]][pos[0]] = "#";
      }
      break;
  }
}

let groundCount = 0;

function fillIn(x, y) {
  let count = 0;
  const queue = [[x, y]];
  while (queue.length) {
    const [x, y] = queue.pop();
    if (x < 0 || x >= width || y < 0 || y >= height) {
      continue;
    }
    if (grid[y][x] === "#") {
      continue;
    }
    if (grid[y][x] === "-") {
      continue;
    }
    if (grid[y][x] === ".") {
      count++;
    }
    grid[y][x] = "-";
    queue.push([x - 1, y]);
    queue.push([x + 1, y]);
    queue.push([x, y - 1]);
    queue.push([x, y + 1]);
  }
  return count;
}

for (let i = 0; i < height; i++) {
  if (grid[i][0] === ".") {
    groundCount += fillIn(0, i);
  }
}

for (let i = 0; i < height; i++) {
  if (grid[i][width - 1] === ".") {
    groundCount += fillIn(width - 1, i);
  }
}

for (let i = 0; i < width; i++) {
  if (grid[0][i] === ".") {
    groundCount += fillIn(i, 0);
  }
}

for (let i = 0; i < width; i++) {
  if (grid[height - 1][i] === ".") {
    groundCount += fillIn(i, height - 1);
  }
}

console.log(height * width - groundCount);

File: 2023/18/solution.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"runtime"
	"strconv"
	"strings"

	"github.com/weitecklee/adventofcode/utils"
)

func main() {
	_, filename, _, _ := runtime.Caller(0)
	dirname := filepath.Dir(filename)
	inputFilePath := filepath.Join(dirname, "input.txt")
	data, err := os.ReadFile(inputFilePath)
	if err != nil {
		panic(err)
	}
	puzzleInput := parseInput(strings.Split(string(data), "\n"))
	fmt.Println(part1(puzzleInput))
	fmt.Println(part2(puzzleInput))
}

var directionConversion = map[byte]byte{
	'0': 'R',
	'1': 'D',
	'2': 'L',
	'3': 'U',
	'R': '0',
	'D': '1',
	'L': '2',
	'U': '3',
}

type Step struct {
	direction byte
	distance  int
	color     string
}

func (s *Step) Swap() *Step {
	distance, err := strconv.ParseInt(s.color[:5], 16, 0)
	if err != nil {
		panic(err)
	}
	color := fmt.Sprintf("%05x%c", s.distance, directionConversion[s.direction])
	return &Step{directionConversion[s.color[5]], int(distance), color}
}

func parseInput(data []string) []*Step {
	steps := make([]*Step, len(data))
	stepRegex := regexp.MustCompile(`(\w) (\d+) \(#(\w{6})\)`)
	for i, line := range data {
		match := stepRegex.FindStringSubmatch(line)
		if match == nil {
			panic(fmt.Sprintf("Could not match regex with: %s", line))
		}
		distance, err := strconv.Atoi(match[2])
		if err != nil {
			panic(err)
		}
		steps[i] = &Step{match[1][0], distance, match[3]}
	}
	return steps
}

func calculateArea(steps []*Step) int {
	// Combination of Shoelace formula and Pick's theorem
	// Shoelace formula calculates area of polygon given vertex coordinates
	// Pick's theorem calculates area of polygon given number of boundary points and number of interior points
	// We have vertices (and boundary points) from input, and ultimately want to find sum of interior points and boundary points
	// Shoelace formula
	// https://en.wikipedia.org/wiki/Shoelace_formula
	// Pick's theorem
	// https://en.wikipedia.org/wiki/Pick%27s_theorem

	var x0, y0, x1, y1, sum, nBoundaryPoints int
	// `x0`, `y0`, `x1`, `y1` are vertex coordinates x_i, y_i, x_i+1, y_i+1,
	// 		with (x0, y0) starting from origin
	// `sum` is sum of (x_i * y_i+1 - x_i+1 * y_i) terms from Shoelace formula
	// `nBoundaryPoints` is count of boundary points

	for _, step := range steps {
		x1, y1 = x0, y0
		switch step.direction {
		case 'U':
			y1 += step.distance
		case 'D':
			y1 -= step.distance
		case 'L':
			x1 -= step.distance
		case 'R':
			x1 += step.distance
		}
		nBoundaryPoints += step.distance
		sum += x0*y1 - x1*y0
		x0, y0 = x1, y1
	}
	area := utils.AbsInt(sum) / 2                   // from Shoelance formula
	nInteriorPoints := area - nBoundaryPoints/2 + 1 // Pick's theorem rearranged
	return nInteriorPoints + nBoundaryPoints
}

func part1(puzzleInput []*Step) int {
	return calculateArea(puzzleInput)
}

func part2(puzzleInput []*Step) int {
	swappedSteps := make([]*Step, len(puzzleInput))
	for i, step := range puzzleInput {
		swappedSteps[i] = step.Swap()
	}
	return calculateArea(swappedSteps)
}
